internal/api/list_stages_with_images_v1alpha1.go

```diff
@@ -0,0 +1,73 @@
+package api
+
+import (
+	"context"
+	"fmt"
+
+	"connectrpc.com/connect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+)
+
+func (s *server) ListStagesWithImages(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.ListStagesWithImagesRequest],
+) (*connect.Response[svcv1alpha1.ListStagesWithImagesResponse], error) {
+	project := req.Msg.GetProject()
+	if err := validateFieldNotEmpty("project", project); err != nil {
+		return nil, err
+	}
+
+	if err := s.validateProjectExists(ctx, project); err != nil {
+		return nil, err
+	}
+
+	var list kargoapi.StageList
+	if err := s.client.List(ctx, &list, client.InNamespace(project)); err != nil {
+		return nil, fmt.Errorf("list stages: %w", err)
+	}
+
+	stages := make([]*kargoapi.Stage, len(list.Items))
+	images := make(map[string]*svcv1alpha1.TagMap)
+
+	for idx, stage := range list.Items {
+		stages[idx] = &list.Items[idx]
+
+		for i, freightGroup := range stage.Status.FreightHistory {
+			for _, freight := range freightGroup.Freight {
+				for _, image := range freight.Images {
```

Comment:
This level of nesting is far from ideal, but barring some larger refactor (like storing images in a top level field of `StageSpec` for example), this is the only option. Previously, very similar logic was performed by the frontend.

Some context makes this less scary:
- Users typically have on the order of 10 Stages per project. Hundreds would be quite rare.
- `stage.Status.FreightHistory` is by default a list of the 10 latest freight. This maximum is not even user-configurable at the moment, so this particular loop can be considered O(1)
- The size of `freightGroup.Freight` corresponds to the number of `requestedFreight` slots in a Stage. This too will typically be a handful of freight at most
- `freight.Images` also rarely has more than a handful of images

In situations where assumptions 1, 3, or 4 are violated, my guess is users would run into _other_ performance bottlenecks before this API becomes a problem